src/util/object_gc_stats.rs

```diff
@@ -0,0 +1,159 @@
+                let obj_lowest_addr = VM::VMObjectModel::object_start_ref(object);
+                if VM::VMObjectModel::HAS_GC_BYTE {
+                    let abs_gc_byte_offset = (object.to_address() - obj_lowest_addr) as isize
+                        + VM::VMObjectModel::GC_BYTE_OFFSET;
+                    // e.g. there is more than 8 bytes from lowest object address to gc byte
+                    if abs_gc_byte_offset >= constants::BYTES_IN_ADDRESS as isize {
+                        obj_lowest_addr // forwarding word at the lowest address of the object storage
+                    } else {
+                        obj_lowest_addr + constants::BYTES_IN_ADDRESS // forwarding word at the first word after the lowest address of the object storage
+                    }
+                } else {
+                    obj_lowest_addr // forwarding word at the lowest address of the object storage
+                }
+            }
+        }
+    }
+
+    #[cfg(target_endian = "big")]
+    fn get_object_status_word_address<VM: VMBinding>(object: ObjectReference) -> Address {
+        unimplemented!()
+    }
+
+    pub fn read<VM: VMBinding>(object: ObjectReference) -> usize {
+        unsafe {
+            Self::get_object_status_word_address::<VM>(object)
+                .atomic_load::<AtomicUsize>(Ordering::SeqCst)
+        }
+    }
+
+    pub fn write<VM: VMBinding>(object: ObjectReference, val: usize) {
+        trace!("GCForwardingWord::write({:#?}, {:x})\n", object, val);
+        unsafe {
+            Self::get_object_status_word_address::<VM>(object)
+                .atomic_store::<AtomicUsize>(val, Ordering::SeqCst)
+        }
+    }
+
+    pub fn compare_exchange<VM: VMBinding>(
+        object: ObjectReference,
+        old: usize,
+        new: usize,
+    ) -> bool {
+        // TODO(Javad): check whether this atomic operation is too strong
+        let res = unsafe {
+            Self::get_object_status_word_address::<VM>(object)
+                .compare_exchange::<AtomicUsize>(old, new, Ordering::SeqCst, Ordering::SeqCst)
+                .is_ok()
+        };
+        trace!(
+            "\nGCForwardingWord::compare_exchange({:#?}, old = {:x}, new = {:x}) -> {}\n",
+            object,
+            old,
+            new,
+            res
+        );
+        res
+    }
+}
+
+// (This function is only used internal to the `util` module)
+//
+// This function checks whether the forwarding word and GC byte can be unified (= the forwarding word fits in the word that contains the GC byte).
+//
+// Returns `None` if the forwarding word and GC byte can not be unified, or if unifiable,
+// returns `Some(fw_offset)`, where `fw_offset` is the offset of the forwarding word relative to `GC_BYTE_OFFSET`.
+//
+// A return value of `Some(fw_offset)` implies that GC byte and forwarding word can be loaded/stored with a single instruction.
+//
+pub(super) fn unifiable_gcbyte_forwarding_word_offset<VM: VMBinding>() -> Option<isize> {
```

Comment:
1. The name you suggested is probably more understandable. And, by unifiable, I mean that forwarding word can be stored in the same word as gc byte.

2. I will slightly change the following sentence in the comment to make it more clear:
From:

`Returns `None` if the forwarding word and GC byte can not be unified, or if unifiable,
// returns `Some(fw_offset)`, where `fw_offset` is the offset of the forwarding word relative to `GC_BYTE_OFFSET`.`

to:

`Returns `None` if the forwarding word and GC byte can not be unified.
Otherwise, returns `Some(fw_offset)`, where `fw_offset` is the offset of the forwarding word relative to `GC_BYTE_OFFSET`.`

Reply:
According to our discussions last week, there should be a minimum object storage size in MMTk, and we never allocate less than that, even if the client VM requests a smaller object storage.
We also decided that the min object storage size is two words.

Considering this, if we solve a scenario where the gc byte is located somewhere in one of these two words, we can solve any other scenarios, because we can always choose a two-word address range in any object, where their gc byte is located.

Reply:
With the assumptions from my previous comment, we can guarantee that one of the following applies to any client VM:

1. gc byte is located in the low-order byte of the first/second word: we can safely do what we are currently doing for jikesrvm.
2. gc byte is located in the high-order byte of the first/second word: we can safely do what we are currently doing for openjdk.
3. gc byte is located in a byte that is not the low/high order byte of any of the two words: we can use the word that does not overlap with the gc byte.

Reply:
> According to our discussions last week, there should be a minimum object storage size in MMTk, and we never allocate less than that, even if the client VM requests a smaller object storage.
> We also decided that the min object storage size is two words.

If we assume the min allocation is two words, we need to have assertions to check this. 
 
> Considering this, if we solve a scenario where the gc byte is located somewhere in one of these two words, we can solve any other scenarios, because we can always choose a two-word address range in any object, where their gc byte is located.

That sounds right.

Reply:
> With the assumptions from my previous comment, we can guarantee that one of the following applies to any client VM:
> 
> 1. gc byte is located in the low-order byte of the first/second word: we can safely do what we are currently doing for jikesrvm.
> 2. gc byte is located in the high-order byte of the first/second word: we can safely do what we are currently doing for openjdk.

This is very VM specific. What makes the highest/lowest byte special? If the gc byte is just an arbitrary byte in the word, why we need to go to case 3. 
Besides, how do you check if the gc byte is in the two words? I don't think there is a VM-neutral way to check that. 

> 3. gc byte is located in a byte that is not the low/high order byte of any of the two words: we can use the word that does not overlap with the gc byte.

Is there a reason that we can't always use this case, use a word that does not include gc byte for forwarding pointers? Then we dont need 1 and 2, and it works for all the cases. Is there any benefit for using the word with gc byte?

Reply:
The special thing about highest/lowest byte is that we can simply `or` the forwarding word's value with the gcbyte's value to achieve the final content of the destination word.
Arbitrary bytes in a word may require partitioning the value of forwarding word to avoid data overwrite.
In other words, if the gcbyte is not the highest/lowest byte of a word, merging it with the forwarding word will require much more calculation that `or`/(`shift`+`or`)

Reply:
The benefit of merging gcbyte+forwarding word is that we only need one load/store operation, compared to two load/store operations when they are in separate words.

Reply:
Yeah. I see.

Reply:
After a discussion with Javad, I suggest doing these:
1. Rename the method to be more clear.
2. Be clear in the comments that only when the gc byte is in the lowest/highest byte in the word, we return `Some` and we use that word for forwarding pointer. For other cases, we use a word in the object for the forwarding pointer. 
3. Add assertions in our allocation code to check min allocation size. 
4. Make it very clear in the comment of our allocation code about the min size, as VM bindings may implement allocation fastpath themselves, and those are unchecked.

Reply:
As clarified in today's meeting:

The min object size is defined by mmtk-core, and transparent to the VM. If the VM allocates something smaller than the min object size, we adjust it to the min size. 

The min object size also depends on the position of gc byte:
1. If there is no gc byte, the min object size is 1 word. We save forwarding pointer in the word. 
2. If the gc byte is low/high order byte, the min object size is 1 word. We save forwarding pointer in the word that contains the gc byte. 
3. For none of the above cases, the min object size is 2 word. We save forwarding pointer in the next word that does not contain the gc byte. 

I feel we could either include this in this PR, or have another PR for the min size after this one.

Reply:
This PR is already too big, lets do that in a separate PR

Reply:
I've done the other things discussed here

Reply:
Sounds good.